test/common/http/http2/metadata_encoder_decoder_test.cc

```diff
@@ -182,6 +182,33 @@ TEST_F(MetadataEncoderDecoderTest, TestMetadataSizeLimit) {
   cleanUp();
 }
 
+TEST_F(MetadataEncoderDecoderTest, TestMetadataSizeLimitIncremental) {
+  MetadataMap metadata_map = {
+      {"header_key1", std::string(1024 * 1024 + 1, 'a')},
+  };
+  MetadataMapPtr metadata_map_ptr = std::make_unique<MetadataMap>(metadata_map);
+  MetadataMapVector metadata_map_vector;
+  metadata_map_vector.push_back(std::move(metadata_map_ptr));
+
+  // Verifies the encoding/decoding result in decoder's callback functions.
+  initialize([this, &metadata_map_vector](MetadataMapPtr&& metadata_map_ptr) -> void {
+    this->verifyMetadataMapVector(metadata_map_vector, std::move(metadata_map_ptr));
+  });
+
+  // metadata_map exceeds size limit.
+  EXPECT_LOG_CONTAINS("error", "exceeds the max bound.",
+                      EXPECT_FALSE(encoder_.createPayload(metadata_map_vector)));
+
+  std::string payload = std::string(1024, 'a');
+  for (int i = 0; i < 1024; ++i) {
+    EXPECT_TRUE(
```

Comment:
I think the "interesting" point in this test is when the payload size is 512 bytes and then another byte is added. The next interesting point is when the payload is 1024 bytes long and a byte is added. I think that loop somewhat obfuscates the issue...
 
 Another, probably clearer way to test this would be to add this test as a friend, and verify `total_payload_size_` directly (or expand the interface to return total payload size).

Reply:
I added an accessor to the interface, thanks.